src/java/org/apache/cassandra/index/sai/disk/PrimaryKeyMap.java

```diff
@@ -61,28 +61,38 @@ default void close() throws IOException
     PrimaryKey primaryKeyFromRowId(long sstableRowId);
 
     /**
-     * Returns a row Id for a {@link PrimaryKey}
+     * Returns a row Id for a {@link PrimaryKey}. If there is no such term, {@link #isNotFound(long)} will return true.
      *
      * @param key the {@link PrimaryKey} to lookup
      * @return the row Id associated with the {@link PrimaryKey}
      */
-    long rowIdFromPrimaryKey(PrimaryKey key);
+    long exactRowIdForPrimaryKey(PrimaryKey key);
 
     /**
-     * Returns the first row Id for a given {@link PrimaryKey}
+     * Returns the sstable row id associated with the least {@link PrimaryKey} greater than or equal to the given
+     * {@link PrimaryKey}. If there is no such term, {@link #isNotFound(long)} will return true.
      *
      * @param key the {@link PrimaryKey} to lookup
      * @return the first row Id associated with the {@link PrimaryKey}
      */
-    long firstRowIdFromPrimaryKey(PrimaryKey key);
+    long ceiling(PrimaryKey key);
 
     /**
-     * Returns the last row Id for a given {@link PrimaryKey}
+     * Returns the sstable row id associated with the greatest {@link PrimaryKey} less than or equal to the given
+     * {@link PrimaryKey}. If there is no such term, {@link #isNotFound(long)} will return true.
      *
      * @param key the {@link PrimaryKey} to lookup
-     * @return the last row Id associated with the {@link PrimaryKey}
+     * @return an sstable row id
      */
-    long lastRowIdFromPrimaryKey(PrimaryKey key);
+    long floor(PrimaryKey key);
+
+    /**
+     * This method is necessary because the two implementations for this interface have divergent behavior for
+     * indicating that a {@link PrimaryKey} was not found.
+     * @param sstableRowId
+     * @return true if the passed row id is not found by the {@link PrimaryKeyMap}
+     */
+    boolean isNotFound(long sstableRowId);
```

Comment:
This feels like a bug waiting to happen again.  Can we standardize the implementations on either negative or MAX_VALUE?  (I would vote for negative)

Reply:
I am happy to standardize. I just wasn't sure how concerned we need to be about backwards compatibility. If we don't, then, I'll fix this now.

Reply:
After a quick look I can't rule out that changing the MAX usages to -1 would break stuff in the trie code

Reply:
what compatibility are you worried about?  it's not worth breaking data on disk for sure, but i don't think this would do that.

Reply:
MAX_VALUE seems to be a construct added in `SortedTermsReader`.

Reply:
Further, `TrieTermsDictionaryReader` already returns `-1` for not found, and the `floor` method used that. It just seems to be generally inconsistent. I'll take another pass.

Reply:
@blambov is there any trie code that implicitly relies on a search for a key returning MAX_VALUE for not found, that would break if we change it to -1 for consistency with LongArray::findTokenRowID and friends?

Reply:
I'll fix in a separate PR

Reply:
This is not something that is defined in the core trie code, it should be completely fine to change. 

While we are there, it's very strange that the SAI code maintains a separate copy of the implementation of `ValueIterator`, and even more so that it would use it to perform exact or ceiling matches. `Walker` itself also provides `followWithGreater` which can be used to implement ceiling in `TrieTermsDictionaryReader` as:
```
    public long ceiling(ByteComparable key)
    {
        int b = followWithGreater(key);
        if (b != ByteSource.END_OF_STREAM)
        {
            if (greaterBranch == -1)
                return NOT_FOUND;
            goMin(greaterBranch);
        }
        
        return getCurrentPayload();
    }
```
(see `PartitionIndex.ceiling`)

Reply:
> This is not something that is defined in the core trie code, it should be completely fine to change.

Thanks for confirming.

> While we are there, it's very strange that the SAI code maintains a separate copy of the implementation of `ValueIterator`, and even more so that it would use it to perform exact or ceiling matches. `Walker` itself also provides `followWithGreater` which can be used to implement ceiling in `TrieTermsDictionaryReader` as:

I was wondering about that code in this thread earlier on the PR with https://github.com/datastax/cassandra/pull/787#discussion_r1347969393 and https://github.com/datastax/cassandra/pull/787#discussion_r1347980879. I didn't have the time/familiarity to say confidently it was unnecessary, but it seems to me that the `TrieRangeIterator` duplicates the `PartitionIndex` logic for ceil and floor while keep track of a stack unnecessarily. Thank you for confirming that the logic seems unnecessarily duplicated. I plan to follow up with a PR to replace usage with something in `TrieTermsDictionaryReader`.

Reply:
> I plan to follow up with a PR to replace usage with something in TrieTermsDictionaryReader

This is exactly what we did in the CEP. The `TrieRangeIterator` is unnecessary because it is only ever used for exact match.